torch/csrc/jit/export.cpp

```diff
@@ -220,19 +220,123 @@ void encodeModel(onnx::ModelProto* p_m, const std::shared_ptr<Graph>& g,
   encodeGraph(p_g, g, initializers);
 }
 
+// Broadcasting operators have the following property:
+// They support a 'broadcast' flag, which enables broadcasting
+// on the last argument.  ATM this is not full-Numpy broadcasting,
+// only left-size extension (no size 1 to size n broadcast)
+std::unordered_set<NodeKind> broadcasting = {
+  kAdd,
+  kDiv,
+  kMul,
+  kPow,
+  kSub,
+  kGemm,
+};
+
+bool isBroadcasting(Node *node) {
+  return broadcasting.count(node->kind());
+}
+
+// When iterating over the dimension sizes, starting at the trailing dimension,
+// the dimension sizes must either be equal, or one of them does not exist.
+//
+//  equivalently:
+//
+// Test that 'from' is a suffix of 'to'.
+bool fusibleExpandTo(at::IntList from, at::IntList to) {
+  auto f = from.rbegin();
+  auto t = to.rbegin();
+  for (; f != from.rend() && t != to.rend(); f++, t++) {
+    // TODO: if 1->n expansion is supported, adjust this conditional.
+    if (*f != *t) return false;
+  }
+  return f == from.rend();
+}
+
+// This optimization fuses expand calls into ONNX operators, because it is
+// easier for non-strided backends to more efficiently do broadcasts if this is
+// local information.  This optimization is not useful for PyTorch as 'expand'
+// is free.
+void fuseBroadcast(const std::shared_ptr<Graph>& graph) {
+  for (auto it = graph->nodes().begin(); it != graph->nodes().end(); ++it) {
+    auto* n = *it;
+
+    // Can't fuse into nodes that don't support broadcasting
+    if (!isBroadcasting(n)) continue;
+
+    // If the node already broadcasts, can't "rebroadcast"
+    // TODO: Actually, maybe you can, if there is a broadcast for some
+    // dims, and then another broadcast for the rest.  But this will
+    // never happen in practice so I didn't implement it.
+    if (n->hasAttribute(kbroadcast) && n->i(kbroadcast)) continue;
+    JIT_ASSERT(!n->hasAttribute(kaxis));
+
+    // TODO: switch ATen tracing to not insert selects for single output.
+    auto* rhs = n->inputs().at(n->inputs().size() - 1);
+
+    // The rhs input isn't actually an expand, so no fusion available
+    if (rhs->kind() != kExpand) continue;
+
+    auto* new_rhs = rhs->input();
+
+    // We need to know what the type pre-expand is.  We should basically
+    // always have this information (because expands are only ever traced,
+    // not generated from symbolic), but if for some reason we don't
+    // have it, we need to skip.
+    if (!new_rhs->hasType()) continue;
+
+    // Not all broadcasts are supported by ONNX broadcast.
+    if (!fusibleExpandTo(new_rhs->type()->expect<TensorType>()->sizes(),    // from
+                         rhs->type()->expect<TensorType>()->sizes()) // to
+       ) continue;
+
+    auto *new_n = graph->createClone(n, [&](Node* n) { return n == rhs ? new_rhs : n; });
```

Comment:
Why not use `replaceInput` on n? It would eliminate all the iterator confusion and `rhs->destroy()` would be safe.